data/is_ocaml_yet/web.md

```diff
@@ -0,0 +1,258 @@
+---
+id: web
+question: Is OCaml web yet?
+answer: Not quite yet, but we're getting there!
+categories:
+  - name: Authentication
+    status: 🆘
+    description: |
+      n/a
+    packages:
+    - FPauth
+    - u2f
+    - webauthn
+    - jwt
+    - cookie
+    - ssl
+    - twostep
+    - oidc
+  - name: Binary protocols
+    status: 🟡
+    description: |
+      There are implementations for a lot of popular protocols, some of them are solid, but some of them are still incomplete and need more testing. Some useful protocols for which there is no package include FlatBuffers and AMQP.
+    packages:
+    - asn1-combinators
+    - avro
+    - capnp
+    - grpc
+    - protobuf
+    - rpc
+    - jsonrpc
+    - thrift
+    - bson
+    - cbor
+    - mqtt
+  - name: Browser
+    status: 🟢
+    description: |
+      OCaml has excellent support for running in the browser!
+      The [Js_of_ocaml](/p/js_of_ocaml/latest) compiler, which translates OCaml into JavaScript, is a well-established tool dating back to 2010. It integrates seamlessly with the existing OCaml ecosystem and powers numerous production applications.
+
+      In addition, [Melange](/p/melange/latest) is a more recent entry to the OCaml-JavaScript compiler space. Originating as a fork of [ReScript](https://rescript-lang.org/), Melange may not have the extensive track record of Js_of_ocaml, but it builds on a very strong foundation and already integrates well with Dune, the OCaml build system.
+    packages:
+    - js_of_ocaml
+    - melange
+    - brr
+    - gen_js_api
+  - name: Compression
+    status: 🟡
+    description: |
+      n/a
+    packages:
+    - zlib
+    - decompress
+    - camlzip
+    - lz4
+    - ezgzip
+    - hpack
+    - brotli
+    - zstd
+    - snappy
+    - lzo
+    - zstandard
+  - name: Crypto
+    status: 🟡
+    description: |
+      n/a
+    packages:
+    - cryptokit
+    - mirage-crypto
+    - nocrypto
+    - pkcs11
+    - tls
+    - hacl_x25519
+    - x509
+    - fiat-p256
+    - sha
+    - blake2
+    - blake3
+    - bls12-381
+    - pbkdf
+    - dirsp-proscript
+    - digestif
+  - name: Database
+    status: 🟠
+    description: |
+      n/a
+    packages:
+    - caqti
+    - pgocaml
+    - sqlite3
+    - lmdb
+    - irmin
+    - mysql
+    - postgresql
+    - aws-rds
+    - sequoia
+    - sqlgg
+    - petrol
+    - mariadb
+  - name: Email
+    status: 🟠
+    description: |
+      n/a
```

Comment:
Also, if you can remove `facteur`, nobody really use it and I will probably archive it soon.

Reply:
Thanks for the suggestions and the description! Would you be able to give some insights on what's missing to get the ideal state for email support in OCaml? Could be gaps in supported protocols, missing libraries, or robustness/maturity. From a developer perspective, what's the story like at the moment if they want to send emails from their web applications?

Reply:
Should this also list ocurl? Curl supports SMTP(S): https://everything.curl.dev/usingcurl/smtp

Reply:
> Would you be able to give some insights on what's missing to get the ideal state for email support in OCaml?

Currently, when it comes to sending emails, we can consider that all the tools are available to a developer. What remains experimental is the implementation of an SMTP service. We can nevertheless extend the existing system with new authentication mechanisms but, as far as I know, those currently available are sufficient.

> Could be gaps in supported protocols, missing libraries, or robustness/maturity.

Robustness & maturity are there. JaneStreet has been using its service for a long time and it's been 3-4 years since I've needed to make a major update to the email sending function - even though this part seems to be used by several web frameworks (like `dream` or `sihl`).

> From a developer perspective, what's the story like at the moment if they want to send emails from their web applications?

I think, still for sending emails, that the main thing is there (generation, protocols, authentication mechanism, verification and signature).

Reply:
> From a developer perspective, what's the story like at the moment if they want to send emails from their web applications?

It depends. 

If you are just sending emails to a reasonably sized group of users or administrators then you can just shell out to `sendmail` like basically any `php` application does (which on debian can be a proxy for something much simpler like [`msmtp`](https://wiki.debian.org/msmtp) if you install `msmtp-mta` that you connect to the mail server of your email provider). So it can be as simple [as that](https://github.com/dbuenzli/hyperbib/blob/a1def764f7f58b29bd732dacaa167e73497f6175/src/email.ml) (modulo the escaping :-).

If you need massive mailing then you likely better off with a mailing service like [mailgun] and then it's simply REST interaction with a webservice. You don't even need to bother with all the standards. The twist here, like with all webservices, is that if you use a popular language the REST interface is likely already implemented for you as a library so it's a bit more work. But are we ready ? Definitively.

[mailgun]: https://www.mailgun.com/